<a href="https://colab.research.google.com/github/rajeshbista/NodeJs_TravisCi_tue/blob/master/Assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 02

Dear Team Members,

I hope you are doing GREAT!

The following assignment is a step-by-step process to implement a multiclass classification using Iris classical dataset.
The idea of the current assignment is to intensify what we have learned in deep learning.

There are some new concepts and challenges that we will face in the current assignment, and we need to resolve each issue and move to the next step.

In week 08 (lecture session), we will discuss part of the potential challenges. The assignment is easy to follow yet needs the patience to solve each question at a time.

**Note:**

Please note that every question has 0.5 and it provides five out of total ten marks for the assignment.

The video presentation, as you have done for Assignment 01, provides the rest of the mark, which is 5.

In short, 5 for coding + 5 for the video presentation = 10.


In [1]:
#00 The iris.csv file (Iris dataset) is located in the Assignment folder below the current ipynb file.

 **Important Note:** We are NOT allowed to import the dataset from Keras. We need to upload the provided iris.csv file to colab from our Google Drive.

In [2]:
#00 1st, upload the iris.csv into your Google Drive (your Colab folder).


In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
#01-a Find the path of iris.csv file, and save it in a variable called fileName.

In [5]:
fileName = "/content/drive/MyDrive/Colab Notebooks/iris.csv"
# from google.colab import files
# uploaded = files.upload()


In [6]:
#01-b Read the content of the iris.csv file and save it in a variable, named iris.

In [7]:
import pandas as pd
import numpy as np
# import io
# iris = pd.read_csv(io.StringIO(uploaded['iris.csv'].decode('utf-8')))
iris = pd.read_csv(fileName)

In [8]:
#01-c Show the first 5 rows of the dataframe.

In [9]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

As you see, the last column contains strings. We want to convert these strings to integers,  and we can work with them easily in our Neural Network model.

setosa -> 0

versicolor -> 1

virginica -> 2

In [10]:
#2-a Please convert the last column strings to integers, as 
# setosa -> 0
# versicolor -> 1
# virginica -> 2

In [11]:
# df['species'] = df['species'].map({'setosa':0, 'versicolor':1, 'virginica':2})
iris.species = pd.factorize(iris.species)[0]


In [12]:
#2-b Now, check the first 5 elements of the iris dataset once again:

In [13]:
iris.head()


sepal_length  sepal_width  petal_length  petal_width  species
0           5.1          3.5           1.4          0.2        0
1           4.9          3.0           1.4          0.2        0
2           4.7          3.2           1.3          0.2        0
3           4.6          3.1           1.5          0.2        0
4           5.0          3.6           1.4          0.2        0

In [14]:
#03-a Select the first four columns of the dataframe as the Data and call the data x.

In [15]:
data = ["sepal_length","sepal_width", "petal_length","petal_width"]
x = iris.loc[:, data]
# x.shape
x.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width'], dtype='object')

In [16]:
#03-b Select the last column as the label and call it y.

In [17]:
y=iris.loc[:,['species']]
y.shape
# y.columns

(150, 1)

In [18]:
#04 Now, it is the time to split the data (x) and the label (y) into training and testin segments, 
#   and call them parts x_train,x_test,y_train,y_test.

In [19]:
#04-a Please split the data, and label ito training and testing segments, while the size of test data is 20%,
#     and use randome_state = 123 to make all the answers similar in the whole class.

In [20]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

In [21]:
#04-b Please, check the first five rows of  x_train,x_test,y_train,y_test.

In [22]:
x_train.head()

sepal_length  sepal_width  petal_length  petal_width
130           7.4          2.8           6.1          1.9
119           6.0          2.2           5.0          1.5
29            4.7          3.2           1.6          0.2
0             5.1          3.5           1.4          0.2
62            6.0          2.2           4.0          1.0

In [23]:
x_test.head()

sepal_length  sepal_width  petal_length  petal_width
72            6.3          2.5           4.9          1.5
112           6.8          3.0           5.5          2.1
132           6.4          2.8           5.6          2.2
88            5.6          3.0           4.1          1.3
37            4.9          3.1           1.5          0.1

In [24]:
y_train.head()

species
130        2
119        2
29         0
0          0
62         1

In [25]:
y_test.head()

species
72         1
112        2
132        2
88         1
37         0

In [26]:
#04-c Check the length of the x_train and x_test. If the numbers are 120 & 30, AWESOME! 
#     You are good to continue!

In [27]:
print("training-length-data:(", x_train.shape, y_train.shape,")")

training-length-data:( (120, 4) (120, 1) )


In [28]:
print("testing-data-length:(", x_test.shape, y_test.shape,")")

testing-data-length:( (30, 4) (30, 1) )


In [29]:
#05-a Please, convert y_train and y_test into categorical, and call them
#     one_hot_train_labels, and one_hot_test_labels, respectively.

In [30]:
from keras.utils import to_categorical
one_hot_train_labels = to_categorical(y_train)
one_hot_test_labels = to_categorical(y_test)

In [31]:
#06-a Please define a sequential model that contains 3 dense layers.
#     In the first and second layers, use 100 hidden units (neurons) and a 'relu' activation function.
#     For the last layer, we need to use 3 as the number of output, 
#     and the reason is we have three classes (setosa -> 0, versicolor -> 1, virginica -> 2)
#     Also, please pick an appropriate activation function for the last layer. 
#     (You already have guessed what could be the best when we have converted the labels to categorical.)

In [32]:
from keras import models
from keras import layers
network = models.Sequential()
network.add(layers.Dense(100, activation='relu'))
network.add(layers.Dense(100, activation='relu'))
network.add(layers.Dense(3, activation='softmax'))


In [33]:
#06-b Now, please compile your model using 'rmsprop', an appropriate loss function, and 'accuracy' for your metrics.

In [34]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

In [35]:
#07-a Pleease fit your model using epoch=20, batch = 10.

In [36]:
network.fit(x_train, one_hot_train_labels, epochs=20, batch_size=10)

Epoch 1/20
12/12 [==============================] - 1s 2ms/step - loss: 0.9444 - accuracy: 0.6147
Epoch 2/20
12/12 [==============================] - 0s 1ms/step - loss: 0.6371 - accuracy: 0.7930
Epoch 3/20
12/12 [==============================] - 0s 2ms/step - loss: 0.5041 - accuracy: 0.8722
Epoch 4/20
12/12 [==============================] - 0s 1ms/step - loss: 0.4362 - accuracy: 0.8920
Epoch 5/20
12/12 [==============================] - 0s 2ms/step - loss: 0.3572 - accuracy: 0.9351
Epoch 6/20
12/12 [==============================] - 0s 1ms/step - loss: 0.3150 - accuracy: 0.9504
Epoch 7/20
12/12 [==============================] - 0s 2ms/step - loss: 0.3070 - accuracy: 0.8854
Epoch 8/20
12/12 [==============================] - 0s 2ms/step - loss: 0.2514 - accuracy: 0.9700
Epoch 9/20
12/12 [==============================] - 0s 2ms/step - loss: 0.2019 - accuracy: 0.9475
Epoch 10/20
12/12 [==============================] - 0s 2ms/step - loss: 0.1878 - accuracy: 0.9541
Epoch 11/20
12/12 [

In [37]:
#08-a Please calculate the result, evaluating your x_test, one_hot_test_label, and see how accurate your model works.

In [38]:
network.evaluate(x_test, one_hot_test_labels)


1/1 [==============================] - 0s 181ms/step - loss: 0.1428 - accuracy: 0.9667


[0.14283248782157898, 0.9666666388511658]

In [39]:
#09-a Now, we repeate all we have done in #06-a with the following parameters:
#     In the first and second layers, use 10 hidden units (neurons) and a 'relu' activation function.
#     For the last layer, we need to use 3. 

In [40]:
network = models.Sequential()
network.add(layers.Dense(10, activation='relu'))
network.add(layers.Dense(10, activation='relu'))
network.add(layers.Dense(3, activation='softmax'))

In [41]:
#09-b Use the same setting as you have done for #06-b (compile phase using 'rmsprop' and et cetera)

In [42]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

In [43]:
#09-c Please fit your model using epoch=4, batch = 10.

In [44]:
network.fit(x_train, one_hot_train_labels, epochs=4, batch_size=10)

Epoch 1/4
12/12 [==============================] - 1s 1ms/step - loss: 2.6927 - accuracy: 0.3493
Epoch 2/4
12/12 [==============================] - 0s 1ms/step - loss: 2.0521 - accuracy: 0.3550
Epoch 3/4
12/12 [==============================] - 0s 1ms/step - loss: 1.6129 - accuracy: 0.3828
Epoch 4/4
12/12 [==============================] - 0s 1ms/step - loss: 1.3786 - accuracy: 0.3791


In [45]:
#09-d Please calculate the result, evaluating your x_test, one_hot_test_label, and see how accurate your model works.

In [46]:
network.evaluate(x_test, one_hot_test_labels)

1/1 [==============================] - 0s 153ms/step - loss: 1.4999 - accuracy: 0.2000


[1.4999024868011475, 0.20000000298023224]

In [47]:
#10-a Please explain all you have learned from this assignment. 
# Theese are some quiestions that you could answer:
# How the number of layers or hidden units could impact the accuracy of our model.
# Do we always suffer from overfitting?
# How can we find and manage overfitting?

Double-click here to start typing your answer: The use of number of hidden layers could have both positive and negative impacts on accuracy, which actually depends more on the complexity of the problem. Based on the above different testing observatond, I can say that increasing sufficient hidden layers definitely improves the output accuracy, while use of much more hidden layers decreses the accuracy.  
Overfitting is common problem in machine learning. By identifying the complexity of problem, we will be able to use only sufficient hidden layers along with number of epochs, so that network generalize well from training data to testing data.

In [48]:
#10-b
# Please try different sessing as we did in #09
# You can change the number of layers, hidden units, compile parameters, number of epochs, batch_size... everything!
# Please change one parameter at a time.
# And share your experience.

Double-click here to start typing your answer: I got the following result from the test upon changing different parameters. 

testdata acc --- traindata acc ----layers # ----layers units ----- epochs 

.93----.96----2----100----20

.93----.97----2----500----20

.96----.98----4----500----40 (accuracy increased in unseen data)

.86----.98----4----100----20(accuracy decreased in unseen data upon less layers and epochs)

.96----.89----4----500----20(accuracy increased in unseen data but decreased in training data upon using less epochs



## Deliverable:
Part A:

Submitting the current colab file.

Part B:

Please record and upload a video to YouTube, describing all you have done in 5 to 10 video recorded clip. If you go beyond 10 minutes, it is fine, and the mentioned time range is to show how the scale of the assignment could be. Then submit a notepad text document that contains the link to the uploaded YouTube video.

